pegen/parser_generator.py

```diff
@@ -163,6 +286,59 @@ def name_loop(self, node: Plain, is_repeat1: bool) -> str:
         return name
 
 
+class PythonParserGenerator(ParserGenerator):
+    def __init__(self, rules: Dict[str, Rule], file: Optional[IO[Text]]):
+        super().__init__(rules, file)
+
+        self.callmakervisitor = PythonCallMakerVisitor(self)
+        self.blockvisitor = PythonBlockVisitor(self)
+        self.bodyvisitor = PythonBodyVisitor(self)
+        self.itemvisitor = PythonItemVisitor(self)
+
+    def pegen_func(self, node):
+        is_loop = node.is_loop()
+        rhs = node.flatten()
+        if node.left_recursive:
+            if node.leader:
+                self.print("@memoize_left_rec")
+            # Non-leader rules in a cycle are not memoized
+        else:
+            self.print("@memoize")
+        self.print(f"def {node.name}(self):")
+        with self.indent():
+            self.print(f"# {node.name}: {rhs}")
+            if node.nullable:
+                self.print(f"# nullable={node.nullable}")
+            self.print("mark = self.mark()")
+            if is_loop:
+                self.print("children = []")
+            self.bodyvisitor.visit(rhs, is_loop=is_loop)
+            if is_loop:
+                self.print("return children")
+            else:
+                self.print("return None")
+
+    def generate_python_module(self, filename: str) -> None:
+        self.print(MODULE_PREFIX.format(filename=filename))
+        self.print("class GeneratedParser(Parser):")
+        while self.todo:
+            for rulename, rule in list(self.todo.items()):
+                del self.todo[rulename]
+                self.print()
+                with self.indent():
+                    self.pegen_func(rule)
+        self.print(MODULE_SUFFIX.rstrip('\n'))
+
+def dedupe(name: str, names: List[str]) -> str:
+    origname = name
+    counter = 0
+    while name in names:
+        counter += 1
+        name = f"{origname}_{counter}"
+    names.append(name)
+    return name
+
+
 def compute_nullables(rules: Dict[str, Rule]) -> None:
```

Comment:
The reason this was not working is because instead of making the parser return a dictionary I am returning a new "Rules" node that have a dictionary inside (the reason is in this way the parser always return a tree of nodes instead of a collection of trees (the rule dictionary)).

This makes things slightly more homogeneous but I can revert that part and just work everywhere with the dictionary. Check 7c8aad2 for the fix, for example.